src/ext/php7/engine_hooks.c

```diff
@@ -547,3 +559,45 @@ void ddtrace_opcode_mshutdown(void) {
 
     zend_set_user_opcode_handler(ZEND_EXIT, NULL);
 }
+
+void ddtrace_execute_internal_minit(void) {
+    _prev_execute_internal = zend_execute_internal;
+    zend_execute_internal = _dd_execute_internal;
+    _dd_call_original_internal = !_prev_execute_internal ? execute_internal : _prev_execute_internal;
+}
+
+void ddtrace_execute_internal_mshutdown(void) { zend_execute_internal = _prev_execute_internal; }
```

Comment:
Think through this with me: we only need one of `_prev_execute_internal` or `_dd_call_original_internal`, I think. If there is a non-null `zend_execute_internal` then we can set prev to that, otherwise set it to `execute_internal`. The only other reason we use this pointer is to restore it in shutdown, which can be done conditionally if the pointer is equal to `execute_internal` or not. Right?

Reply:
Good point. I think that'll be cleaner. :)

Reply:
I think so too. Having two pointers always felt a bit weird before, but in all the times I've played with this I only realized this now lol.

Reply:
Yeah, it hadn't occurred to me either! :D